File: src/main/java/me/pcasaes/hexoids/infrastructure/kafka/KafkaBoltLifeCycleConsumerRebalanceListener.java
package me.pcasaes.hexoids.infrastructure.kafka;

import io.smallrye.reactive.messaging.kafka.KafkaConsumerRebalanceListener;
import org.apache.kafka.clients.consumer.Consumer;
import org.eclipse.microprofile.config.inject.ConfigProperty;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.inject.Named;
import java.util.Collection;
import java.util.logging.Logger;
import java.util.stream.Collectors;

@ApplicationScoped
@Named("hexoids-server-bolt-life-cycle")
public class KafkaBoltLifeCycleConsumerRebalanceListener implements KafkaConsumerRebalanceListener {

    private static final Logger LOGGER = Logger.getLogger(KafkaBoltLifeCycleConsumerRebalanceListener.class.getName());

    private final DelayedStartConsumerHandler handler;
    private final int boltMaxDuration;

    @Inject
    public KafkaBoltLifeCycleConsumerRebalanceListener(
            DelayedStartConsumerHandler handler,
            @ConfigProperty(
                    name = "hexoids.config.bolt.max.duration",
                    defaultValue = "10000"
            ) int boltMaxDuration) {
        this.handler = handler;
        this.boltMaxDuration = boltMaxDuration;
    }

    @Override
    public void onPartitionsAssigned(Consumer<?, ?> consumer, Collection<org.apache.kafka.common.TopicPartition> partitions) {
        long now = System.currentTimeMillis();
        long shouldStartAt = now - (this.boltMaxDuration + 10_000L);

        consumer.offsetsForTimes(partitions
                .stream()
                .collect(Collectors.toMap(k -> k, v -> shouldStartAt))
        )
                .forEach((k, v) -> {
                            LOGGER.info("Seeking to " + v);
                            consumer.seek(k, v.offset());
                            LOGGER.info("Seeked to " + v);
                        }
                );

        handler
                .onPartitionsAssigned()
                .await()
                .indefinitely();
    }


}

File: src/main/kotlin/me/pcasaes/hexoids/infrastructure/kafka/KafkaBoltLifeCycleConsumerRebalanceListener.kt
package me.pcasaes.hexoids.infrastructure.kafka

import io.smallrye.reactive.messaging.kafka.KafkaConsumerRebalanceListener
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.inject.Named
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.common.TopicPartition
import org.eclipse.microprofile.config.inject.ConfigProperty
import java.util.logging.Logger

@ApplicationScoped
@Named("hexoids-server-bolt-life-cycle")
class KafkaBoltLifeCycleConsumerRebalanceListener @Inject constructor(
    private val handler: DelayedStartConsumerHandler,
    @param:ConfigProperty(
        name = "hexoids.config.bolt.max.duration",
        defaultValue = "10000"
    ) private val boltMaxDuration: Int
) : KafkaConsumerRebalanceListener {

    override fun onPartitionsAssigned(consumer: Consumer<*, *>, partitions: MutableCollection<TopicPartition>) {
        val now = System.currentTimeMillis()
        val shouldStartAt = now - (this.boltMaxDuration + 10_000L)

        consumer.offsetsForTimes(
            partitions.associateWith { shouldStartAt }
        ).forEach { (k, v) ->
            if (v != null) {
                LOGGER.info("Seeking to $v")
                consumer.seek(k, v.offset())
                LOGGER.info("Seeked to $v")
            }
        }

        handler
            .onStarted()
            .await()
            .indefinitely()
    }


    companion object {
        private val LOGGER: Logger = Logger.getLogger(KafkaBoltLifeCycleConsumerRebalanceListener::class.java.getName())
    }
}
